File: slider/src/main/java/me/aungkooo/slider/ViewAdapter.java
package me.aungkooo.slider;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v4.view.PagerAdapter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Ko Oo on 25/5/2018.
 */

public abstract class ViewAdapter<V extends View, OBJ> extends PagerAdapter {
    private final Context context;
    private List<OBJ> itemList;
    private V view;

    public ViewAdapter(Context context) {
        this.context = context;
        this.itemList = new ArrayList<>();
    }

    public ViewAdapter(Context context, List<OBJ> itemList) {
        this.context = context;
        this.itemList = itemList;
    }

    public void setItemList(ArrayList<OBJ> itemList) {
        this.itemList = itemList;
        notifyDataSetChanged();
    }

    public List<OBJ> getItemList() {
        return itemList;
    }

    public OBJ getItem(int position) {
        return itemList.get(position);
    }

    public Context getContext() {
        return context;
    }

    public V getView() {
        return view;
    }

    @SuppressWarnings("unchecked")
    public V createView(@LayoutRes int resId, ViewGroup container) {
        return (V) LayoutInflater.from(context).inflate(resId, container, false);
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        view = onCreateView(container);
        final OBJ item = getItem(position);
        onBindView(container, view, item, position);
        container.addView(view);
        return view;
    }

    @Override
    public int getCount() {
        return itemList.size();
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
        return view == object;
    }

    @Override
    public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
        container.removeView((View) object);
    }

    public abstract V onCreateView(ViewGroup container);

    public abstract void onBindView(ViewGroup container, final V view, final OBJ item, int position);
}

File: slider/src/main/java/me/aungkooo/slider/ViewAdapter.kt
package me.aungkooo.slider

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.viewpager.widget.PagerAdapter
import java.util.*

/**
 * Created by Ko Oo on 25/5/2018.
 */
@Suppress("unused", "MemberVisibilityCanBePrivate")
abstract class ViewAdapter<V : View?, OBJ> : PagerAdapter {
    val context: Context
    var itemList: List<OBJ>
        private set
    var view: V? = null
        private set

    constructor(context: Context) {
        this.context = context
        itemList = ArrayList()
    }

    constructor(context: Context, itemList: List<OBJ>) {
        this.context = context
        this.itemList = itemList
    }

    fun setItemList(itemList: ArrayList<OBJ>) {
        this.itemList = itemList
        notifyDataSetChanged()
    }

    fun getItem(position: Int): OBJ {
        return itemList[position]
    }

    fun createView(@LayoutRes resId: Int, container: ViewGroup?): V {
        @Suppress("UNCHECKED_CAST")
        return LayoutInflater.from(context).inflate(resId, container, false) as V
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        view = onCreateView(container)
        val item = getItem(position)
        onBindView(container, view, item, position)
        container.addView(view)
        return view as Any
    }

    override fun getCount(): Int {
        return itemList.size
    }

    override fun isViewFromObject(view: View, `object`: Any): Boolean {
        return view === `object`
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as View)
    }

    abstract fun onCreateView(container: ViewGroup?): V
    abstract fun onBindView(container: ViewGroup?, view: V?, item: OBJ, position: Int)
}